### PR TITLE
Dms endpoint module

### PR DESCRIPTION
**Pull request by ruimoreira**
##### SUMMARY
Added new module for AWS Data migration service endpoints

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
dms_endpoint
##### ADDITIONAL INFORMATION
This module can be used to manage

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/dms_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+            want to use to create the endpoint.
+    externaltabledefinition:
+       description:
+         - The external table definition
+    dynamodbsettings:
+       description:
+         - Settings in JSON format for the target Amazon DynamoDB endpoint
+           if source or target is dynamodb
+    s3settings:
+       description
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/dms_endpoint.py]: This is your problem, you're missing a colon. I had to load the whole thing using `yaml.load` to spot it

```
>>> DOCUMENTATION = '''...
>>> import yaml
>>> docs = yaml.load(DOCUMENTATION)
>>> for opt in docs['options']:
...     print opt, type(docs['options'][opt])
... 
username <type 'dict'>
elasticsearchsettings <type 'dict'>
servername <type 'dict'>
dynamodbsettings <type 'dict'>
serviceaccessrolearn <type 'dict'>
kmskeyid <type 'dict'>
dmstransfersettings <type 'dict'>
enginename <type 'dict'>
password <type 'dict'>
endpointidentifier <type 'dict'>
port <type 'dict'>
wait <type 'dict'>
certificatearn <type 'dict'>
tags <type 'dict'>
endpointtype <type 'dict'>
state <type 'dict'>
externaltabledefinition <type 'dict'>
extraconnectionattributes <type 'dict'>
databasename <type 'dict'>
sslmode <type 'dict'>
mongodbsettings <type 'dict'>
kinesissettings <type 'dict'>
s3settings <type 'str'>
```

Comment (ruimoreira) [reply lib/ansible/modules/cloud/amazon/dms_endpoint.py]: @willthames thank you for your help.

**Comment by willthames** [comment]
@ruimoreira for now you can use unsupported rather than aws/groupX in the aliases - that will make the automated tests pass until the permissions are added - we can run the tests manually in the interim.

@mattclay / @s-hertel can you let @ruimoreira know if there's anything to be done to get automated tests working?

File: test/integration/targets/dms_endpoint/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,135 @@
+---
+
+- name: set connection information for all tasks
+  set_fact:
+    aws_connection_info: &aws_connection_info
+      aws_access_key: "{{ aws_access_key }}"
+      aws_secret_key: "{{ aws_secret_key }}"
+      #security_token: "{{ security_token }}"
+      region: "{{ aws_region }}"
+  no_log: no
```
Comment (BobBoldin) [inline test/integration/targets/dms_endpoint/tasks/main.yml]: please set no_log: yes

Comment (ruimoreira) [reply test/integration/targets/dms_endpoint/tasks/main.yml]: Done ! Thank you

**Comment by BobBoldin** [comment]
Ran tests locally, all passed though expect on parallel testing will cross-detect resources.  Will add comment to test file with possible update.  Likely should have an always section to block in case a test fails it will still delete the endpoint.

ansible 2.9.0.dev0
python version = 2.7.5
Red Hat Enterprise Linux Server release 7.5 (Maipo)

For CI permissions, testing output:
AWS ACTIONS: ['dms:CreateEndpoint', 'dms:DeleteEndpoint', 'dms:DescribeEndpoints', 'dms:ModifyEndpoint']

File: test/integration/targets/dms_endpoint/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+- name: set connection information for all tasks
+  set_fact:
+    aws_connection_info: &aws_connection_info
+      aws_access_key: "{{ aws_access_key }}"
+      aws_secret_key: "{{ aws_secret_key }}"
+      #security_token: "{{ security_token }}"
+      region: "{{ aws_region }}"
+  no_log: yes
+
```
Comment (BobBoldin) [inline test/integration/targets/dms_endpoint/tasks/main.yml:11]: May want to create a fact so the endpointidentifier will always be unique.


- name: generate unique value for testing
  set_fact:
    dms_identifier: "{{ resource_prefix }}-dms"

File: test/integration/targets/dms_endpoint/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+- block:
+  - name: create endpoints
+    dms_endpoint:
+        state: present
+        endpointidentifier: 'testsource'
```
Comment (BobBoldin) [inline test/integration/targets/dms_endpoint/tasks/main.yml]: change reference to fact if decide to use it.
```suggestion
        endpointidentifier: "{{ dms_identifier }}"
```

File: test/integration/targets/dms_endpoint/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+  - assert:
+      that:
+        - result is changed
+        - result is not failed
+
+  - name: create endpoints no change
+    dms_endpoint:
+        state: present
+        endpointidentifier: 'testsource'
```
Comment (BobBoldin) [inline test/integration/targets/dms_endpoint/tasks/main.yml]: Same and all other occurrences update to use fact identifier

File: test/integration/targets/dms_endpoint/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      databasename: 'testdb'
+      sslmode: none
+      wait: False
+      <<: *aws_connection_info
+    register: result
+
+  - assert:
+      that:
+        - result is not changed
+        - result is not failed
```
Comment (BobBoldin) [inline test/integration/targets/dms_endpoint/tasks/main.yml]:
```suggestion
        - result is not failed
 always:
   ###### TEARDOWN STARTS HERE ######
   - name: delete endpoints
    dms_endpoint:
      state: absent
      endpointidentifier: 'testsource'
      endpointtype: source
      enginename: aurora
      username: testing
      password: testint1234
      servername: testing.moreira.dom
      port: 3306
      databasename: 'testdb'
      sslmode: none
      wait: True
      timeout: 60
      retries: 10
      <<: *aws_connection_info
   
```

**Review by BobBoldin — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
Thanks @BobBoldin for the additional suggestions, those were pretty much my remaining concerns too.

Once those are addressed this should be good to merge

**Comment by ruimoreira** [comment]
Will commit this asap.
Thank you both for your help

**Comment by mattclay** [comment]
I've added this PR to the [backlog](https://github.com/ansible/ansible/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+label%3Aneeds_ci_update) of permissions updates for our CI system to support automated testing. Unfortunately I don't know when we'll be able to get to this. Marking the test `unsupported` in the meantime will allow merging -- at the risk of having future modifications to the module and/or tests potentially result in undetected failures.

**Comment by ruimoreira** [comment]
@willthames @BobBoldin I have done the suggested changes, please review. 
I will now start working on another PR for another module still on the DMS side of AWS

**Comment by ruimoreira** [comment]
Can someone have a last look at this ? 
Should be ready for merge according to feedback provided

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
Merged, thanks @ruimoreira

**Pull request merged**